Failed to link with multiple definition of X

# Problem

ngx_mruby fails to link with multiple definition of X errors when 
-fno-common is enabled.

Since gcc version 10, -fno-common is enabled by default, so multiple definition of X error occurs.
It seems that it is  well known issue [1] on Fedora 32, for example.

[1] https://src.fedoraproject.org/rpms/redhat-rpm-config/blob/master/f/buildflags.md#legacy-fcommon

It needs to eliminate multiple definition correctly. (the following code may fix one of them)

```
diff --git a/src/http/ngx_http_mruby_core.c b/src/http/ngx_http_mruby_core.c
index 9a2cd33..f647e37 100644
--- a/src/http/ngx_http_mruby_core.c
+++ b/src/http/ngx_http_mruby_core.c
@@ -13,7 +13,7 @@
 #include "mruby/string.h"
 #include "mruby/variable.h"
 
-ngx_module_t ngx_http_mruby_module;
+extern ngx_module_t ngx_http_mruby_module;
 
 #if (NGX_DEBUG)
 static void ngx_mrb_log_backtrace(mrb_state *mrb, mrb_value obj, ngx_log_t *log)
diff --git a/src/stream/ngx_stream_mruby_core.c b/src/stream/ngx_stream_mruby_core.c
index 5c619ea..deea022 100644
--- a/src/stream/ngx_stream_mruby_core.c
+++ b/src/stream/ngx_stream_mruby_core.c
@@ -10,7 +10,7 @@
 
 #include <mruby/hash.h>
 #include <mruby/string.h>
-ngx_module_t ngx_stream_mruby_module;
+extern ngx_module_t ngx_stream_mruby_module;
 static mrb_value ngx_stream_mrb_errlogger(mrb_state *mrb, mrb_value self)
 {
   mrb_value msg;
```

# Expected

No linker error with -fno-common.


# How to reproduce

Step 1. Enable `-fno-common` flag
Step 2. Run ./test.sh

In step 1, you can enable -fno-common like this:

```
diff --git a/test.sh b/test.sh
index 027a02b..7ebd6fb 100755
--- a/test.sh
+++ b/test.sh
@@ -18,7 +18,7 @@ if [ `uname -s` = "NetBSD" ]; then
     PS_C="pgrep -l"
 else
     NPROCESSORS_ONLN="_NPROCESSORS_ONLN"
-    NGINX_DEFAULT_OPT='--with-debug --with-http_stub_status_module --with-http_ssl_module --with-cc-opt=-g\ -O0'
+    NGINX_DEFAULT_OPT='--with-debug --with-http_stub_status_module --with-http_ssl_module --with-cc-opt=-g\ -O0\ -fno-common'
     MAKE=make
     if [ -f /etc/centos-release ]; then
         KILLALL=pkill
```

# Actual

```
cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -g -O0 -fno-common -I src/core -I src/event -I src/event/modules -I src/os/unix -I /work/nginx/ngx_mruby/mruby/src -I /work/nginx/ngx_mruby/mruby/include -I -std=gnu99 -I -g -I -O3 -I -Wall -I -Wundef -I -Wdeclaration-after-statement -I -Werror-implicit-function-declaration -I -Wwrite-strings -I -I/work/nginx/ngx_mruby/mruby/include -I -I/work/nginx/ngx_mruby/mruby/build/repos/host/mruby-pointer/include -I -I/usr/local/include -I /work/nginx/ngx_mruby/dependence/ngx_devel_kit/objs -I objs/addon/ndk -I objs \
	-o objs/ngx_modules.o \
	objs/ngx_modules.c
cc -o objs/nginx \
objs/src/core/nginx.o \
objs/src/core/ngx_log.o \
objs/src/core/ngx_palloc.o \
objs/src/core/ngx_array.o \
objs/src/core/ngx_list.o \
objs/src/core/ngx_hash.o \
objs/src/core/ngx_buf.o \
objs/src/core/ngx_queue.o \
objs/src/core/ngx_output_chain.o \
objs/src/core/ngx_string.o \
objs/src/core/ngx_parse.o \
objs/src/core/ngx_parse_time.o \
objs/src/core/ngx_inet.o \
objs/src/core/ngx_file.o \
objs/src/core/ngx_crc32.o \
objs/src/core/ngx_murmurhash.o \
objs/src/core/ngx_md5.o \
objs/src/core/ngx_sha1.o \
objs/src/core/ngx_rbtree.o \
objs/src/core/ngx_radix_tree.o \
objs/src/core/ngx_slab.o \
objs/src/core/ngx_times.o \
objs/src/core/ngx_shmtx.o \
objs/src/core/ngx_connection.o \
objs/src/core/ngx_cycle.o \
objs/src/core/ngx_spinlock.o \
objs/src/core/ngx_rwlock.o \
objs/src/core/ngx_cpuinfo.o \
objs/src/core/ngx_conf_file.o \
objs/src/core/ngx_module.o \
objs/src/core/ngx_resolver.o \
objs/src/core/ngx_open_file_cache.o \
objs/src/core/ngx_crypt.o \
objs/src/core/ngx_proxy_protocol.o \
objs/src/core/ngx_syslog.o \
objs/src/event/ngx_event.o \
objs/src/event/ngx_event_timer.o \
objs/src/event/ngx_event_posted.o \
objs/src/event/ngx_event_accept.o \
objs/src/event/ngx_event_udp.o \
objs/src/event/ngx_event_connect.o \
objs/src/event/ngx_event_pipe.o \
objs/src/os/unix/ngx_time.o \
objs/src/os/unix/ngx_errno.o \
objs/src/os/unix/ngx_alloc.o \
objs/src/os/unix/ngx_files.o \
objs/src/os/unix/ngx_socket.o \
objs/src/os/unix/ngx_recv.o \
objs/src/os/unix/ngx_readv_chain.o \
objs/src/os/unix/ngx_udp_recv.o \
objs/src/os/unix/ngx_send.o \
objs/src/os/unix/ngx_writev_chain.o \
objs/src/os/unix/ngx_udp_send.o \
objs/src/os/unix/ngx_udp_sendmsg_chain.o \
objs/src/os/unix/ngx_channel.o \
objs/src/os/unix/ngx_shmem.o \
objs/src/os/unix/ngx_process.o \
objs/src/os/unix/ngx_daemon.o \
objs/src/os/unix/ngx_setaffinity.o \
objs/src/os/unix/ngx_setproctitle.o \
objs/src/os/unix/ngx_posix_init.o \
objs/src/os/unix/ngx_user.o \
objs/src/os/unix/ngx_dlopen.o \
objs/src/os/unix/ngx_process_cycle.o \
objs/src/os/unix/ngx_linux_init.o \
objs/src/event/modules/ngx_epoll_module.o \
objs/src/os/unix/ngx_linux_sendfile_chain.o \
objs/src/event/ngx_event_openssl.o \
objs/src/event/ngx_event_openssl_stapling.o \
objs/src/core/ngx_regex.o \
objs/src/http/ngx_http.o \
objs/src/http/ngx_http_core_module.o \
objs/src/http/ngx_http_special_response.o \
objs/src/http/ngx_http_request.o \
objs/src/http/ngx_http_parse.o \
objs/src/http/modules/ngx_http_log_module.o \
objs/src/http/ngx_http_request_body.o \
objs/src/http/ngx_http_variables.o \
objs/src/http/ngx_http_script.o \
objs/src/http/ngx_http_upstream.o \
objs/src/http/ngx_http_upstream_round_robin.o \
objs/src/http/ngx_http_file_cache.o \
objs/src/http/ngx_http_write_filter_module.o \
objs/src/http/ngx_http_header_filter_module.o \
objs/src/http/modules/ngx_http_chunked_filter_module.o \
objs/src/http/modules/ngx_http_range_filter_module.o \
objs/src/http/modules/ngx_http_gzip_filter_module.o \
objs/src/http/ngx_http_postpone_filter_module.o \
objs/src/http/modules/ngx_http_ssi_filter_module.o \
objs/src/http/modules/ngx_http_charset_filter_module.o \
objs/src/http/modules/ngx_http_userid_filter_module.o \
objs/src/http/modules/ngx_http_headers_filter_module.o \
objs/src/http/ngx_http_copy_filter_module.o \
objs/src/http/modules/ngx_http_not_modified_filter_module.o \
objs/src/http/modules/ngx_http_static_module.o \
objs/src/http/modules/ngx_http_autoindex_module.o \
objs/src/http/modules/ngx_http_index_module.o \
objs/src/http/modules/ngx_http_mirror_module.o \
objs/src/http/modules/ngx_http_try_files_module.o \
objs/src/http/modules/ngx_http_auth_basic_module.o \
objs/src/http/modules/ngx_http_access_module.o \
objs/src/http/modules/ngx_http_limit_conn_module.o \
objs/src/http/modules/ngx_http_limit_req_module.o \
objs/src/http/modules/ngx_http_geo_module.o \
objs/src/http/modules/ngx_http_map_module.o \
objs/src/http/modules/ngx_http_split_clients_module.o \
objs/src/http/modules/ngx_http_referer_module.o \
objs/src/http/modules/ngx_http_rewrite_module.o \
objs/src/http/modules/ngx_http_ssl_module.o \
objs/src/http/modules/ngx_http_proxy_module.o \
objs/src/http/modules/ngx_http_fastcgi_module.o \
objs/src/http/modules/ngx_http_uwsgi_module.o \
objs/src/http/modules/ngx_http_scgi_module.o \
objs/src/http/modules/ngx_http_memcached_module.o \
objs/src/http/modules/ngx_http_empty_gif_module.o \
objs/src/http/modules/ngx_http_browser_module.o \
objs/src/http/modules/ngx_http_upstream_hash_module.o \
objs/src/http/modules/ngx_http_upstream_ip_hash_module.o \
objs/src/http/modules/ngx_http_upstream_least_conn_module.o \
objs/src/http/modules/ngx_http_upstream_random_module.o \
objs/src/http/modules/ngx_http_upstream_keepalive_module.o \
objs/src/http/modules/ngx_http_upstream_zone_module.o \
objs/src/http/modules/ngx_http_stub_status_module.o \
objs/src/stream/ngx_stream.o \
objs/src/stream/ngx_stream_variables.o \
objs/src/stream/ngx_stream_script.o \
objs/src/stream/ngx_stream_handler.o \
objs/src/stream/ngx_stream_core_module.o \
objs/src/stream/ngx_stream_log_module.o \
objs/src/stream/ngx_stream_proxy_module.o \
objs/src/stream/ngx_stream_upstream.o \
objs/src/stream/ngx_stream_upstream_round_robin.o \
objs/src/stream/ngx_stream_write_filter_module.o \
objs/src/stream/ngx_stream_limit_conn_module.o \
objs/src/stream/ngx_stream_access_module.o \
objs/src/stream/ngx_stream_geo_module.o \
objs/src/stream/ngx_stream_map_module.o \
objs/src/stream/ngx_stream_split_clients_module.o \
objs/src/stream/ngx_stream_return_module.o \
objs/src/stream/ngx_stream_upstream_hash_module.o \
objs/src/stream/ngx_stream_upstream_least_conn_module.o \
objs/src/stream/ngx_stream_upstream_random_module.o \
objs/src/stream/ngx_stream_upstream_zone_module.o \
objs/addon/http/ngx_http_mruby_module.o \
objs/addon/http/ngx_http_mruby_core.o \
objs/addon/http/ngx_http_mruby_init.o \
objs/addon/http/ngx_http_mruby_request.o \
objs/addon/http/ngx_http_mruby_var.o \
objs/addon/http/ngx_http_mruby_connection.o \
objs/addon/http/ngx_http_mruby_server.o \
objs/addon/http/ngx_http_mruby_filter.o \
objs/addon/http/ngx_http_mruby_upstream.o \
objs/addon/http/ngx_http_mruby_ssl.o \
objs/addon/http/ngx_http_mruby_async.o \
objs/addon/stream/ngx_stream_mruby_module.o \
objs/addon/stream/ngx_stream_mruby_init.o \
objs/addon/stream/ngx_stream_mruby_core.o \
objs/addon/stream/ngx_stream_mruby_connection.o \
objs/addon/stream/ngx_stream_mruby_async.o \
objs/addon/src/ndk.o \
objs/ngx_modules.o \
-ldl -lpthread -lcrypt -L/work/nginx/ngx_mruby/mruby/build/host/lib -lmruby -lm -L/work/nginx/ngx_mruby/mruby/build/host/lib -L/usr/local/lib -lmruby -lm -lreadline -lncurses -lcrypto -lpthread -lpthread -lrt -lpthread -lrt /work/nginx/ngx_mruby/mruby/build/host/mrbgems/mruby-redis/hiredis/lib/libhiredis.a -L -lmruby -lpcre -lssl -lcrypto -ldl -lpthread -lz \
-Wl,-E
/bin/ld: objs/addon/http/ngx_http_mruby_core.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:106: multiple definition of `ngx_http_next_header_filter'; objs/addon/http/ngx_http_mruby_module.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:106: first defined here
/bin/ld: objs/addon/http/ngx_http_mruby_core.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:107: multiple definition of `ngx_http_next_body_filter'; objs/addon/http/ngx_http_mruby_module.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:107: first defined here
/bin/ld: objs/addon/http/ngx_http_mruby_core.o:(.bss+0x20): multiple definition of `ngx_http_mruby_module'; objs/addon/http/ngx_http_mruby_module.o:(.data.rel.local+0x40): first defined here
/bin/ld: objs/addon/http/ngx_http_mruby_var.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:106: multiple definition of `ngx_http_next_header_filter'; objs/addon/http/ngx_http_mruby_module.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:106: first defined here
/bin/ld: objs/addon/http/ngx_http_mruby_var.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:107: multiple definition of `ngx_http_next_body_filter'; objs/addon/http/ngx_http_mruby_module.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:107: first defined here
/bin/ld: objs/addon/http/ngx_http_mruby_server.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:106: multiple definition of `ngx_http_next_header_filter'; objs/addon/http/ngx_http_mruby_module.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:106: first defined here
/bin/ld: objs/addon/http/ngx_http_mruby_server.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:107: multiple definition of `ngx_http_next_body_filter'; objs/addon/http/ngx_http_mruby_module.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:107: first defined here
/bin/ld: objs/addon/http/ngx_http_mruby_upstream.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:106: multiple definition of `ngx_http_next_header_filter'; objs/addon/http/ngx_http_mruby_module.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:106: first defined here
/bin/ld: objs/addon/http/ngx_http_mruby_upstream.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:107: multiple definition of `ngx_http_next_body_filter'; objs/addon/http/ngx_http_mruby_module.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:107: first defined here
/bin/ld: objs/addon/http/ngx_http_mruby_ssl.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:106: multiple definition of `ngx_http_next_header_filter'; objs/addon/http/ngx_http_mruby_module.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:106: first defined here
/bin/ld: objs/addon/http/ngx_http_mruby_ssl.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:107: multiple definition of `ngx_http_next_body_filter'; objs/addon/http/ngx_http_mruby_module.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:107: first defined here
/bin/ld: objs/addon/http/ngx_http_mruby_async.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:106: multiple definition of `ngx_http_next_header_filter'; objs/addon/http/ngx_http_mruby_module.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:106: first defined here
/bin/ld: objs/addon/http/ngx_http_mruby_async.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:107: multiple definition of `ngx_http_next_body_filter'; objs/addon/http/ngx_http_mruby_module.o:/work/nginx/ngx_mruby/src/http/ngx_http_mruby_module.h:107: first defined here
/bin/ld: objs/addon/stream/ngx_stream_mruby_core.o:/work/nginx/ngx_mruby/src/stream/ngx_stream_mruby_core.c:13: multiple definition of `ngx_stream_mruby_module'; objs/addon/stream/ngx_stream_mruby_module.o:/work/nginx/ngx_mruby/src/stream/ngx_stream_mruby_module.c:117: first defined here
collect2: error: ld returned 1 exit status
```

--- a/build.sh
+++ b/build.sh
@@ -16,11 +16,11 @@ set -e
 # OS specific configuration
 if [ `uname -s` = "NetBSD" ]; then
     NPROCESSORS_ONLN="NPROCESSORS_ONLN"
-    NGINX_DEFAULT_OPT='--with-http_stub_status_module --with-stream --without-stream_access_module --with-ld-opt=-L/usr/pkg/lib\ -Wl,-R/usr/pkg/lib'
+    NGINX_DEFAULT_OPT='--with-http_stub_status_module --with-stream --without-stream_access_module --with-cc-opt=-fno-common --with-ld-opt=-L/usr/pkg/lib\ -Wl,-R/usr/pkg/lib'
     MAKE=gmake
 else
     NPROCESSORS_ONLN="_NPROCESSORS_ONLN"
-    NGINX_DEFAULT_OPT='--with-http_stub_status_module --with-stream --without-stream_access_module'
+    NGINX_DEFAULT_OPT='--with-http_stub_status_module --with-stream --without-stream_access_module --with-cc-opt=-fno-common'
     MAKE=make
 fi
 

--- a/src/http/ngx_http_mruby_core.c
+++ b/src/http/ngx_http_mruby_core.c
@@ -13,8 +13,6 @@
 #include "mruby/string.h"
 #include "mruby/variable.h"
 
-ngx_module_t ngx_http_mruby_module;
-
 #if (NGX_DEBUG)
 static void ngx_mrb_log_backtrace(mrb_state *mrb, mrb_value obj, ngx_log_t *log)
 {

--- a/src/http/ngx_http_mruby_module.c
+++ b/src/http/ngx_http_mruby_module.c
@@ -27,6 +27,9 @@
     (code)->ctx = NULL;                                                                                                \
   }
 
+ngx_http_output_header_filter_pt ngx_http_next_header_filter;
+ngx_http_output_body_filter_pt ngx_http_next_body_filter;
+
 // set conf
 static void *ngx_http_mruby_create_srv_conf(ngx_conf_t *cf);
 static char *ngx_http_mruby_merge_srv_conf(ngx_conf_t *cf, void *parent, void *child);

--- a/src/http/ngx_http_mruby_module.h
+++ b/src/http/ngx_http_mruby_module.h
@@ -103,7 +103,7 @@ typedef struct ngx_http_mruby_loc_conf_t {
   ngx_http_output_body_filter_pt body_filter_handler;
 } ngx_http_mruby_loc_conf_t;
 
-ngx_http_output_header_filter_pt ngx_http_next_header_filter;
-ngx_http_output_body_filter_pt ngx_http_next_body_filter;
+extern ngx_http_output_header_filter_pt ngx_http_next_header_filter;
+extern ngx_http_output_body_filter_pt ngx_http_next_body_filter;
 
 #endif // NGX_HTTP_MRUBY_MODULE_H

--- a/src/stream/ngx_stream_mruby_core.c
+++ b/src/stream/ngx_stream_mruby_core.c
@@ -10,7 +10,7 @@
 
 #include <mruby/hash.h>
 #include <mruby/string.h>
-ngx_module_t ngx_stream_mruby_module;
+
 static mrb_value ngx_stream_mrb_errlogger(mrb_state *mrb, mrb_value self)
 {
   mrb_value msg;

--- a/src/stream/ngx_stream_mruby_core.h
+++ b/src/stream/ngx_stream_mruby_core.h
@@ -23,4 +23,6 @@ typedef struct ngx_stream_mruby_ctx_t {
 
 ngx_stream_mruby_ctx_t *ngx_stream_mrb_get_module_ctx(mrb_state *mrb, ngx_stream_session_t *s);
 
+extern ngx_module_t ngx_stream_mruby_module;
+
 #endif // NGX_STREAM_MRUBY_CORE_H

--- a/test.sh
+++ b/test.sh
@@ -12,13 +12,13 @@ set -e
 # OS specific configuration
 if [ `uname -s` = "NetBSD" ]; then
     NPROCESSORS_ONLN="NPROCESSORS_ONLN"
-    NGINX_DEFAULT_OPT='--with-debug --with-http_stub_status_module --with-http_ssl_module --with-ld-opt=-L/usr/pkg/lib\ -Wl,-R/usr/pkg/lib --with-cc-opt=-g\ -O0'
+    NGINX_DEFAULT_OPT='--with-debug --with-http_stub_status_module --with-http_ssl_module --with-ld-opt=-L/usr/pkg/lib\ -Wl,-R/usr/pkg/lib --with-cc-opt=-g\ -O0\ -fno-common'
     MAKE=gmake
     KILLALL=pkill
     PS_C="pgrep -l"
 else
     NPROCESSORS_ONLN="_NPROCESSORS_ONLN"
-    NGINX_DEFAULT_OPT='--with-debug --with-http_stub_status_module --with-http_ssl_module --with-cc-opt=-g\ -O0'
+    NGINX_DEFAULT_OPT='--with-debug --with-http_stub_status_module --with-http_ssl_module --with-cc-opt=-g\ -O0\ -fno-common'
     MAKE=make
     if [ -f /etc/centos-release ]; then
         KILLALL=pkill

